In [0]:
%sql
SELECT * FROM workspace.bronze_layer_staging.stg_customers

customer_id first_name last_name email phone_number address city country postal_code membership_status last_updated 17138 Cayla Vaudrey cvaudrey0@mozilla.com 934-100-4597 3626 Memorial Terrace Kertapura Indonesia null inactive 2011-06-03 15424 Brant Owttrim bowttrim1@ebay.co.uk 297-506-9475 4 Waxwing Lane Algoz Portugal 8365-059 active 2011-05-20 17364 Grayce Calderon gcalderon2@bloglines.com 989-992-9054 4 Jackson Place Luhyny Ukraine null inactive 2010-09-07 17712 Artie Tuting atuting3@washingtonpost.com 775-776-1447 38 Farwell Alley Chelopech Bulgaria 2087 inactive 2010-11-20 12447 Godart Carlone gcarlone4@joomla.org 260-927-6971 39 Larry Terrace Jiabeiyan China null pending 2010-12-12 14632 Dorisa Ney dney5@jimdo.com 914-809-6493 52733 Village Green Hill Gayam Indonesia null inactive 2011-03-23 15952 Sheelah Hedge shedge6@vimeo.com 654-205-7392 23 Dapin Center Mondoteko Indonesia null active 2011-05-12 17723 Beverlie Beseke bbeseke7@netvibes.com 233-971-3024 7 Welch Crossing Witbank South Africa 2250 active 2011-04-20 15668 Morry Geany mgeany8@myspace.com 394-428-0190 2537 Huxley Lane Dobratice Czech Republic 739 51 active 2010-05-07 14354 Giacopo Adenot gadenot9@uol.com.br 454-576-2827 76 Ludington Parkway ‘Ābūd Palestinian Territory null active 2010-08-01 12871 Sascha Fyndon sfyndona@slideshare.net 246-977-9598 80468 Sheridan Pass Longtang China null pending 2011-11-11 16430 Raymund Cleveley rcleveleyb@thetimes.co.uk 442-226-6638 337 Oxford Center Planovskoye Russia 361222 inactive 2010-02-15 13310 Babbette Floyed bfloyedc@timesonline.co.uk 919-116-5226 0830 Memorial Street Raleigh United States 27690 inactive 2010-07-20 16170 Lamar Keighley lkeighleyd@intel.com 627-219-4716 07 Lerdahl Point Talon Indonesia null inactive 2011-05-11 15782 Chlo Peyto cpeytoe@amazon.co.jp 192-417-0608 823 Pierstorff Trail Kalenderovci Donji Bosnia and Herzegovina null active 2011-05-11 17618 Nerita Reffe nreffef@businessinsider.com 681-890-9714 61626 Ilene Drive Hondo Japan 503-0232 active 2011-03-19 12363 Jamaal Beurich jbeurichg@guardian.co.uk 686-900-9405 3 Carpenter Point Sarzedo Brazil 32450-000 pending 2011-05-31 15756 Lorraine Agutter lagutterh@over-blog.com 552-333-7135 39 Colorado Center Cigarogol Indonesia null pending 2010-01-31 18016 Ninon Mapstone nmapstonei@imgur.com 180-476-2768 9891 Cottonwood Street Belo Cameroon null pending 2011-08-01 15348 Diena Bussons dbussonsj@google.it 578-589-5770 11071 Crowley Parkway Banayoyo Philippines 2708 pending 2011-08-24 16563 Brendis Basilotta bbasilottak@time.com 269-592-3954 30 Granby Hill Ar Rābiyah Kuwait null inactive 2011-12-08 14733 Gizela Fordyce gfordycel@about.me 948-772-1474 0109 Jenna Lane Villa del Rosario Colombia 541038 pending 2011-07-03 13673 Cammy Snelson csnelsonm@delicious.com 512-580-0387 9 Goodland Terrace Moshkovo Russia 633130 pending 2010-11-17 13139 Sherwynd Molines smolinesn@arizona.edu 925-645-5534 0 Hoepker Parkway Centre de Flacq Mauritius null active 2011-03-27 13011 Shannon Heaselgrave sheaselgraveo@webeden.co.uk 159-635-8585 45 Sloan Park Sylvan Lake Canada T4S inactive 2010-02-09 13596 Lothario Blakeley lblakeleyp@examiner.com 564-337-3379 02 Messerschmidt Hill São Pedro Portugal 2710-470 pending 2011-11-14 15027 Herman Speedin hspeedinq@netvibes.com 774-663-8095 02 Starling Avenue Mīrwāh Gorchani Pakistan 36121 pending 2010-05-24 17176 Gwenette Kerwen gkerwenr@seattletimes.com 841-900-1673 179 Ridgeview Street Lubasz Poland 64-720 pending 2011-08-27 13240 Cletis Gilley cgilleys@foxnews.com 601-252-8342 17 Sachs Way Araulí Honduras null pending 2011-11-25 13555 Lemuel Dahl ldahlt@deviantart.com 971-108-8783 8 Monument Point Placencia Belize null pending 2010-12-23 17720 Tedman Baudain tbaudainu@skype.com 318-570-0444 83 Roxbury Drive Shreveport United States 71105 pending 2011-02-13 17609 Debra O'Glassane doglassanev@drupal.org 780-684-7753 62 Merchant Center Moa Cuba null pending 2011-11-04 15789 Osborn Gogin ogoginw@dmoz.org 506-399-2198 874 Jana Alley C

In [0]:
%sql
SELECT COUNT(distinct customer_id) FROM workspace.bronze_layer_staging.stg_customers

count(DISTINCT customer_id) 2711

### Exploring the Relationship Between Customer ID and Email

Let's examine how customer IDs are associated with email addresses in our dataset.

In [0]:
%sql
SELECT email, COUNT(customer_id) FROM workspace.bronze_layer_staging.stg_customers
GROUP BY email
HAVING COUNT(customer_id) > 1

email count(customer_id)

In [0]:
%sql
SELECT customer_id, COUNT(email) FROM workspace.bronze_layer_staging.stg_customers
GROUP BY customer_id
HAVING COUNT(email) > 1

customer_id count(email) 17138 2 15424 2 17364 3 15952 4 14354 2 13310 2 16170 2 15782 3 17618 2 15756 2 16563 2 13011 2 13596 2 13240 3 13555 3 17720 2 17609 2 15789 2 12952 2 14076 2 16759 2 17325 3 15219 4 17651 2 14032 3 15732 2 13506 3 18015 2 14219 3 15597 2 12702 2 13300 3 16762 2 15363 2 13570 2 13050 2 17829 4 17017 2 15312 3 16143 2 15993 4 13265 4 15110 4 17416 2 16036 3 13486 3 15396 4 14562 3 14005 2 14951 2 13675 3 13347 2 17912 2 16275 2 13008 2 16320 2 15510 3 14619 2 14649 4 13565 3 15160 2 14396 2 13947 3 16955 2 14463 2 13259 2 16598 2 15059 4 12752 2 14626 2 12837 2 12583 2 12494 2 14049 2 14768 3 12561 2 16115 2 12950 2 14920 2 15153 2 18263 2 13751 5 16574 2 13668 2 15423 2 17788 2 14179 5 13525 3 16570 2 13850 2 14981 2 15149 2 12888 3 17457 2 14173 3 13972 3 13560 2 16091 2 12412 2 18237 4 13449 2 15722 2 17643 2 12374 2 14165 2 17898 2 17410 3 16017 2 12514 4 17135 2 12715 2 13561 2 17535 4 14528 4 14013 2 14034 2 17629 2 18035 3 17382 2 14627 2 15491 2 15081 2 18122 3 14447 2 17864 3 15332 2 17987 3 13823 2 17068 2 16934 2 14703 2 17747 3 16076 2 13634 2 14535 2 15195 2 17118 2 16720 2 16672 2 12669 3 16936 4 13203 3 13509 3 15038 4 16835 2 17131 2 16513 2 16050 2 14592 2 17758 2 12817 2 13647 4 16164 2 12532 3 12558 3 15005 2 15462 3 16658 2 12609 2 18143 2 15734 2 13508 3 15255 2 16701 2 16960 2 14661 2 13485 2 14307 2 16871 3 16161 3 16560 2 15339 3 15350 2 17866 2 17159 2 15773 2 12883 2 14964 2 15275 2 15676 3 12719 2 17905 3 15159 2 15582 2 16708 2 13109 2 15381 2 16293 2 14215 2 13956 2 17865 2 16175 3 13848 3 17790 2 14087 4 12651 3 14971 3 17279 2 17824 2 13348 3 13924 3 15823 4 15215 3 15076 2 13396 3 18118 2 16581 3 17913 2 13592 2 17962 3 12454 3 15970 2 16995 2 17653 2 14198 2 16098 2 14287 3 16022 2 12749 3 17370 2 17806 2 16905 2 14422 2 16203 2 13792 2 16042 2 15822 2 14027 3 12507 2 12536 2 12868 2 15150 2 16561 2 15223 3 14500 3 16226 4 14739 3 13846 3 15360 2 13716 2 16684 3 14748 2 17229 4 13470 3 14400 2 15695 5 17154 3 13155 2 14970 2 15733 3 12380 3 14132 2 13502 2 17250 3 15839 2 12491 2 15932 2 14221 3 12989 3 13458 3 16116 2 16557 2 14240 2 15232 3 14622 2 16686 3 13887 2 15744 2 14375 3 14256 2 15122 2 17734 2 14140 2 14957 2 13892 2 14348 2 17828 2 15922 3 16957 2 15910 2 15827 4 17509 4 15019 2 17569 3 14416 4 14882 2 18074 3 12957 5 13727 2 15574 3 13573 2 13590 2 16412 2 13130 2 15400 2 16693 3 16669 3 17750 2 14518 2 14902 3 12503 2 13982 2 12659 2 17377 2 17698 2 14910 2 12700 2 13001 2 15626 3 17088 2 13473 2 18136 2 13548 2 17858 2 13726 2 13365 3 16769 2 12585 2 12757 2 15216 3 14676 2 17514 2 12626 2 17862 2 15355 2 12473 3 17621 2 16601 2 13594 3 12971 2 17777 3 14536 2 15468 2 17561 2 12426 3 16360 2 13282 2 17331 2 18257 2 15201 2 12879 2 16225 4 15095 3 17442 2 15953 2 12856 5 12623 3 17820 2 14860 2 16766 3 17724 4 14713 3 17453 2 16968 3 14393 2 14388 3 12601 2 12947 4 16676 2 13375 2 12935 2 14155 2 14159 3 16353 2 16038 2 16893 3 15639 2 13730 2 14868 4 16187 2 14247 2 16579 2 16754 3 14054 3 15124 3 14972 2 13075 3 16018 3 14261 4 16303 2 15471 3 12640 2 12924 2 13568 2 14127 3 17890 2 14367 2 16440 2 12820 2 13629 2 15428 2 17946 2 14923 2 15589 2 17042 2 14660 2 13217 2 15175 2 13484 2 13969 2 13552 2 14948 2 15045 2 17409 2 16119 2 14290 3 15569 2 16109 3 17505 2 12383 2 13748 2 12657 2 15738 3 16756 2 13029 2 16650 2 12792 3 16054 2 13703 2 14125 2 15607 2 15193 2 13566 2 16173 2 16830 2 16450 2 12753 2 13806 2 13040 2 14030 2 12425 2 12527 2 12851 2 16451 2 13658 2 16505 2 15212 2 14647 2 16052 3 13822 2 17924 2 14723 2 16859 4 14725 3 14199 2 17589 3 16696 2 14659 2 16084 2 16848 2 14553 2 13844 4 13227 2 15083 3 16327 4 14538 2 13705 2 14785 2 13419 3 15795 2 18133 2 15603 2 15570 2 15031 2 17547 2 18096 3 16751 2 17974 3 12607 2 13082 2 16595 2 16207 2 15449 2 14850 2 15894 2 14507 3 14851 3 15706 2 14498 2 14395 3 15834 2 15165 2 16020 3 12417 2 12373 3 13865 2 15018 2 16048 3 15965 2 16239 2 14497 2 14022 2 16711 2 13026 3 12988 2 16843 3 12398 2

The results above indicate that some customer IDs are linked to multiple email addresses. This is generally considered poor data hygiene and may require further investigation.Lets pick one such case and investigate it.

In [0]:
%sql
SELECT * FROM  workspace.bronze_layer_staging.stg_customers
WHERE customer_id = '15952'

customer_id first_name last_name email phone_number address city country postal_code membership_status last_updated 15952 Sheelah Hedge shedge6@vimeo.com 654-205-7392 23 Dapin Center Mondoteko Indonesia null active 2011-05-12 15952 Brandise Rohan brohanav@jimdo.com 144-741-3692 15 Westridge Place Le Hochet Mauritius null inactive 2010-02-18 15952 Myrah Lainton mlaintong6@miibeian.gov.cn 241-638-1422 235 Jenifer Street Ivanishchi Russia 601521 pending 2010-08-26 15952 Wendeline Berdale wberdalef9@narod.ru 460-637-9412 139 Fair Oaks Way Graz Austria 8047 pending 2011-08-05

Let's focus on customers with an **active** membership status. We'll create a Silver layer model to gain insights into these customers. If there are multiple active records for the same customer ID, we'll select the most recently updated one.

In [0]:
%sql
WITH deduplicated_customers AS (
  SELECT *
  FROM (
    SELECT *,
      row_number() OVER (
        PARTITION BY customer_id
        ORDER BY last_updated DESC
      ) AS latest_rec
    FROM workspace.bronze_layer_staging.stg_customers
    WHERE customer_id IN (
      SELECT customer_id
      FROM workspace.bronze_layer_staging.stg_customers
      GROUP BY customer_id
      HAVING COUNT(email) > 1
    )
    ORDER BY customer_id
  )
  WHERE latest_rec = 1
)
SELECT
  customer_id,
  first_name,
  last_name,
  email,
  phone_number,
  address,
  city,
  country,
  postal_code,
  membership_status,
  last_updated
FROM
  deduplicated_customers
UNION
SELECT * FROM workspace.bronze_layer_staging.stg_customers
    WHERE customer_id IN (
      SELECT customer_id
      FROM workspace.bronze_layer_staging.stg_customers
      GROUP BY customer_id
      HAVING COUNT(email) = 1
    )

customer_id first_name last_name email phone_number address city country postal_code membership_status last_updated 12507 Dietrich McNevin dmcnevinb8@jigsy.com 426-802-6500 3 Butternut Junction Lodan Wetan Indonesia null active 2011-09-22 12649 Daron Kirwood dkirwoodfh@hp.com 689-475-4897 67757 Emmet Lane Huanchillas Argentina 6121 inactive 2011-08-19 12791 Alaric Bathoe abathoe32@geocities.jp 621-793-4232 366 Steensland Plaza Mubo China null pending 2011-10-12 12815 Inna Everington ieveringtone2@icio.us 780-538-7674 5 Lien Street Antananarivo Madagascar null pending 2010-11-12 12879 Ardyth Arton aartonrh@gravatar.com 424-605-2427 43596 Derek Drive Alagoa Grande Brazil 58388-000 pending 2011-06-04 12942 Egbert Steabler esteablerb3@time.com 923-920-6797 42939 1st Lane Sanxi China null active 2011-04-30 13050 Julianne Gytesham jgyteshamfk@godaddy.com 680-961-9742 6 Helena Place Syanno Belarus null active 2011-06-13 13079 Melly Josef mjosef5k@amazonaws.com 147-720-8128 8463 Mockingbird Lane Pesek Indonesia null pending 2010-12-22 13300 Sheree Tyson stysongw@youku.com 319-369-1194 13 Vernon Court Cedar Rapids United States 52405 inactive 2011-07-17 13310 Isidoro Ridgwell iridgwelllj@hubpages.com 875-125-2248 30 Summit Park Jinsha China null inactive 2011-02-19 13365 Dari Deason ddeason1c@1688.com 634-112-0219 9361 School Hill Qixian China null active 2011-12-17 13373 Rock Adrian radrianp1@ifeng.com 101-609-6664 9 Banding Road Norrköping Sweden 603 86 pending 2011-10-18 13473 Denyse Stainer dstainerjy@plala.or.jp 494-542-3561 4 Tony Park Marapat Indonesia null active 2010-09-21 13596 Lothario Blakeley lblakeleyp@examiner.com 564-337-3379 02 Messerschmidt Hill São Pedro Portugal 2710-470 pending 2011-11-14 13716 Vivianne Cunniff vcunniff9i@1und1.de 532-249-2530 13439 Badeau Avenue Qo’ng’irot Shahri Uzbekistan null active 2011-10-03 13722 Nicky Vivash nvivashhy@theatlantic.com 236-329-3761 7 Banding Court Flor da Mata Portugal 2840-707 pending 2011-11-11 13771 Nancie Astman nastmanmh@ning.com 733-860-5921 762 Monterey Place Dinjo Indonesia null active 2011-06-21 13865 Daniel Kermath dkermath69@un.org 222-462-7231 7510 Burning Wood Parkway Fatumuti Indonesia null active 2011-03-07 13887 Fayth Howford fhowfordcw@berkeley.edu 959-809-4643 8 Lyons Alley Nowe Brzesko Poland 32-120 pending 2011-08-24 13950 Major Bartoszek mbartoszek32@1und1.de 451-352-9086 658 Springview Drive Bum Bum Venezuela null pending 2010-08-07 14199 Alwyn Burfield aburfieldnu@bravesites.com 921-835-3144 05 Alpine Street Purral Costa Rica 10807 active 2011-02-09 14336 Marcela Tugwell mtugwellfi@dion.ne.jp 670-397-0584 54 Mendota Hill Strel'na Russia 301735 active 2011-04-15 14344 Valentine Vannah vvannaho2@dell.com 443-137-6178 9 Tony Way Xiangxi China null active 2011-07-21 14348 Daniela Feaks dfeaksgs@scribd.com 596-535-1202 70491 Luster Parkway Likhoy Russia 356501 pending 2010-11-25 14354 Giacopo Adenot gadenot9@uol.com.br 454-576-2827 76 Ludington Parkway ‘Ābūd Palestinian Territory null active 2010-08-01 14404 Sephira Posnett sposnettqs@toplist.cz 386-265-2499 9 Colorado Trail Perico Cuba null active 2010-12-25 14496 Huntley MacCostigan hmaccostiganai@unicef.org 654-736-2029 8 Saint Paul Circle Tashtagol Russia 652993 inactive 2011-12-01 14524 Shermy Haldenby shaldenbyke@washingtonpost.com 825-778-1054 147 Paget Park Villaviciosa Philippines 2811 active 2011-03-07 14538 Lemmy Spafford lspafford6@ask.com 608-730-8599 24977 Basil Alley Garunggang Indonesia null inactive 2011-01-11 14562 Matthaeus Ferreo mferreo2j@bandcamp.com 975-823-9883 02724 Vermont Court Timezgana Morocco null active 2011-03-24 14619 Willi Roches wroches26@bing.com 542-986-2350 9275 Jenna Junction Červenka Czech Republic 784 01 pending 2011-12-05 14823 Nissa Pietroni npietronikr@csmonitor.com 442-705-0212 40966 Rockefeller Parkway Ancasti Argentina 4701 pending 2011-08-15 14932 Dorine Baff dbaffda@blogtalkradio.com 969-631-2226 20383 Bartelt Circle Borovo Bulgaria 7119 active 2011-07-24 1495

The final count of records in the customers table, after removing duplicates, will be **2711**.

Let's validate this result to confirm that only duplicate records are removed and no extra records are lost.

## Count Validation

In [0]:
%sql
--original number of records in table
SELECT COUNT(*) FROM workspace.bronze_layer_staging.stg_customers

count(1) 4000

In [0]:
%sql
-- total number of records where customer_id is associated with multiple emails
SELECT 
  SUM(email_count) AS total_email_count
FROM (
  SELECT 
    customer_id, 
    COUNT(email) AS email_count
  FROM workspace.bronze_layer_staging.stg_customers
  GROUP BY customer_id
  HAVING COUNT(email) > 1
)

total_email_count 2301

In [0]:
%sql
-- total number of customer ids with multiple email
SELECT 
  COUNT(*) AS num_customer_ids_with_multiple_emails
FROM (
  SELECT customer_id
  FROM workspace.bronze_layer_staging.stg_customers
  GROUP BY customer_id
  HAVING COUNT(email) > 1
)

num_customer_ids_with_multiple_emails 1012


### Deduplication Validation

From the previous two cells, we observe:

- **2,301** records have multiple email issues
- Affecting **1,012** unique customer IDs
- Out of a total of **4,000** original records

---

**After applying our fix, the expected number of records is:**

\[
(4,000 - 2,301) + 1,012 = 2,711
\]

---

This confirms that our deduplication process is correct and **no extra records are lost**.